sdk/servicebus/Azure.Messaging.ServiceBus/samples/Sample15_MockingClientTypes.md

```diff
@@ -0,0 +1,1347 @@
+    .Returns(mockProcessor.Object);
+
+// Start the processor.
+
+await mockProcessor.Object.StartProcessingAsync();
+
+// This is where application code would be called and tested.
+
+// Stop the processor.
+
+await mockProcessor.Object.StopProcessingAsync();
+```
+
+### For the `ServiceBusSessionProcessor`
+
+```C# Snippet:ServiceBus_SimulateRunningTheSessionProcessor
+Mock<ServiceBusClient> mockClient = new();
+Mock<ServiceBusSessionReceiver> mockSessionReceiver = new();
+
+// This handler is for illustrative purposes only.
+
+async Task MessageHandler(ProcessSessionMessageEventArgs args)
+{
+    string body = args.Message.Body.ToString();
+    Console.WriteLine(body);
+
+    // we can evaluate application logic and use that to determine how to settle the message.
+    await args.CompleteMessageAsync(args.Message);
+}
+
+// This function simulates a random message being emitted for processing.
+
+Random rng = new();
+
+TimerCallback dispatchMessage = async _ =>
+{
+    ServiceBusReceivedMessage message = ServiceBusModelFactory.ServiceBusReceivedMessage(
+        body: new BinaryData("message"),
+        messageId: "messageId",
+        sessionId: "session",
+        partitionKey: "hellokey",
+        correlationId: "correlationId",
+        contentType: "contentType",
+        replyTo: "replyTo"
+        //...
+        );
+
+    ProcessSessionMessageEventArgs processArgs = new(
+        message: message,
+        receiver: mockSessionReceiver.Object,
+        cancellationToken: CancellationToken.None);
+
+    await MessageHandler(processArgs);
+};
+
+// Create a timer that runs once-a-second when started and, otherwise, sits idle.
+
+Timer messageDispatchTimer = new(
+    dispatchMessage,
+    null,
+    Timeout.Infinite,
+    Timeout.Infinite);
+
+void startTimer() =>
+    messageDispatchTimer.Change(0, (int)TimeSpan.FromSeconds(1).TotalMilliseconds);
+
+void stopTimer() =>
+    messageDispatchTimer.Change(Timeout.Infinite, Timeout.Infinite);
+
+// Create a mock of the session processor that dispatches messages when StartProcessingAsync
+// is called and does so until StopProcessingAsync is called.
+
+Mock<ServiceBusSessionProcessor> mockProcessor = new();
+
+mockProcessor
+    .Setup(processor => processor.StartProcessingAsync(It.IsAny<CancellationToken>()))
+    .Returns(Task.CompletedTask)
+    .Callback(startTimer);
+
+mockProcessor
+    .Setup(processor => processor.StopProcessingAsync(It.IsAny<CancellationToken>()))
+    .Returns(Task.CompletedTask)
+    .Callback(stopTimer);
+
+mockClient
+    .Setup(client => client.CreateSessionProcessor(It.IsAny<string>(), It.IsAny<ServiceBusSessionProcessorOptions>()))
+    .Returns(mockProcessor.Object);
+
+// Start the processor.
+
+await mockProcessor.Object.StartProcessingAsync();
+
+// This is where application code would be called and tested.
+
+// Stop the processor.
+
+await mockProcessor.Object.StopProcessingAsync();
+```
+
+## Sending and Receiving messages using topics and subscriptions
```

Comment:
Consider: splitting this into two snippets one for sending and one for receiving

Reply:
... and one for CRUD?    😄